File: problems/ts/8.stack-qeueue/validate-string.ts
export function validateString(txt: string) {
    let stack = [];

    for (let i = 0; i < txt.length; ++i) {
        let char = txt[i];

        if (char == '(') {
            stack.push([char, i]);
            continue;
        }

        if (char == ')' && stack.length == 0) {
            txt = txt.substring(0, i) + txt.substring(i + 1, txt.length);
            i--;
            continue;
        }

        if (char == ')' && stack.length > 0) {
            stack.pop();
        }
    }

    for (let i = 0; i < stack.length; i++) {
        let [_, index] = stack.pop() as [string, number];
        i--;
        txt = txt.substring(0, +index) + txt.substring(+index + 1, txt.length);
    }

    return txt;
}

File: golang/5.stack-qeueue/validate_string.go
package main

func validateString(txt string) string {
	stack := []int{}

	for i := 0; i < len(txt); i++ {
		char := txt[i]

		if char == '(' {
			stack = append(stack, i)
			continue
		}

		if char == ')' && len(stack) == 0 {
			txt = txt[:i] + txt[i+1:]
			i -= 1
			continue
		}

		if char == ')' && len(stack) > 0 {
			stack = stack[:len(stack)-1]
		}
	}

	for i := 0; i < len(stack); i++ {
		removeIndex := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		i--
		txt = txt[:removeIndex] + txt[removeIndex+1:]
	}

	return txt
}
